File: 0713-subarray-product-less-than-k/0713-subarray-product-less-than-k.cpp
class Solution
{
public:
    int numSubarrayProductLessThanK(vector<int> &nums, int k)
    {
        if(k<=1) return 0;
        int ans = 0;
        int start = 0, end = 0;
        int mulResult = 1;
        while (end < nums.size())
        {
            mulResult *= nums[end];

            while (mulResult >= k)
            {
                mulResult = mulResult / nums[start];
                start++;
            }

            ans += end - start + 1;
            end++;
        }
        return ans;
    }
};

File: 0713-subarray-product-less-than-k/0713-subarray-product-less-than-k.java
class Solution {
    public int numSubarrayProductLessThanK(int[] nums, int k) {

        int res = 0;
        int left = 0;
        int product = 1;
        for (int i = 0; i < nums.length; i++) {
            product *= nums[i];

            while (product >= k && left <=i) {
                product/=nums[left];
                left++;
            }

            int addedSubArrays = i-left+1;
            // because each time you add a new number
            // you are not only add one sub array, but it may be more than one sub array
            res+=addedSubArrays;
        }
        return res;
    }
}